src/irc/core/irc-servers.c

```diff
@@ -968,6 +985,46 @@ void irc_server_init_isupport(IRC_SERVER_REC *server)
 	}
 }
 
+void event_cap_start (IRC_SERVER_REC *server)
+{
+	if (gslist_find_string(server->cap, "sasl")) {
+		g_warning("Got sasl in the list");
+		irc_send_cmd_now(server, "CAP REQ :sasl");
+	}
+}
+
+void event_auth (IRC_SERVER_REC *server, const char *data)
+{
+	char *blob, *enc;
+	int len;
+	IRC_SERVER_CONNECT_REC *conn;
+
+	conn = server->connrec;
+
+	if (conn->password == NULL || data[0] != '+') {
+		/* abort the authentication */
+		irc_send_cmd_now(server, "AUTHENTICATE *");
+		cap_register_done(server);
+		return;
+	}
+
+	blob = g_strdup_printf("%s-%s-%s",
+												 conn->username,
+												 conn->username,
+												 conn->password);
+	len = strlen(blob);
+
+	blob[strlen(conn->username)] = '\0';
+	blob[strlen(conn->username) * 2 + 1] = '\0';
```

Comment:
Since you're dealing with strings that have NULs in them, maybe it would be saner to use [GString][]s?

[Gstring]: https://developer.gnome.org/glib/stable/glib-Strings.html